Compatibility issues in unzLocateFile

Hi,

I've run into a building issue with the `libdigidocpp` package when switching from minizip (madler/zlib part) to minizip-ng (built with MZ_COMPAT flag).

I reported this to their upstream [1], but now when I think about it, it may be the issue of minizip-ng as it should be fully compatible, but it's not.

Could you please take a look at the error and help me resolve this building issue? I was trying to find some sort of PORTING document in your Github, but there is nothing that would look like it.
Thank you

[1] https://github.com/open-eid/libdigidocpp/issues/569

I've added the `#include <zlib.h>` to the `ZipSerialize.cpp` file, as they were counting on minizip to include it.

But now I've encountered another error, which may be the incompatible part of minizip-ng function:
Minizip-ng definition: `unzLocateFile(unzFile file, const char *filename, unzFileNameComparer filename_compare_func)`
Minizip definition: `unzLocateFile OF((unzFile file, const char *szFileName, int iCaseSensitivity))`

Error:
```
/builddir/build/BUILD/libdigidocpp-3.16.0/src/util/ZipSerialize.cpp:139:58: error: invalid conversion from ‘int’ to ‘unzFileNameComparer’ {aka ‘int (*)(void*, const char*, const char*)’} [-fpermissive]
  139 |     int unzResult = unzLocateFile(d->open, file.c_str(), 1);
      |                                                          ^
      |                                                          |
      |                                                          int
In file included from /usr/include/minizip/unzip.h:10,
                 from /builddir/build/BUILD/libdigidocpp-3.16.0/src/util/ZipSerialize.cpp:27:
/usr/include/minizip/mz_compat.h:352:87: note:   initializing argument 3 of ‘int unzLocateFile(unzFile, const char*, unzFileNameComparer)’
  352 | ZEXPORT int     unzLocateFile(unzFile file, const char *filename, unzFileNameComparer filename_compare_func);
      |                                                                   ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/libdigidocpp-3.16.0/src/util/ZipSerialize.cpp: In member function ‘void digidoc::ZipSerialize::addFile(const std::string&, std::istream&, const Properties&, Flags)’:
/builddir/build/BUILD/libdigidocpp-3.16.0/src/util/ZipSerialize.cpp:189:17: error: cannot convert ‘<brace-enclosed initializer list>’ to ‘uint32_t’ {aka ‘unsigned int’} in initialization
  189 |         0, 0, 0 };
      |                 ^
/builddir/build/BUILD/libdigidocpp-3.16.0/src/util/ZipSerialize.cpp: In member function ‘digidoc::ZipSerialize::Properties digidoc::ZipSerialize::properties(const std::string&) const’:
/builddir/build/BUILD/libdigidocpp-3.16.0/src/util/ZipSerialize.cpp:225:58: error: invalid conversion from ‘int’ to ‘unzFileNameComparer’ {aka ‘int (*)(void*, const char*, const char*)’} [-fpermissive]
  225 |     int unzResult = unzLocateFile(d->open, file.c_str(), 1);
      |                                                          ^
      |                                                          |
      |                                                          int
/usr/include/minizip/mz_compat.h:352:87: note:   initializing argument 3 of ‘int unzLocateFile(unzFile, const char*, unzFileNameComparer)’
  352 | ZEXPORT int     unzLocateFile(unzFile file, const char *filename, unzFileNameComparer filename_compare_func);
      |                                                                   ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
gmake[2]: *** [src/CMakeFiles/digidocpp.dir/build.make:489: src/CMakeFiles/digidocpp.dir/util/ZipSerialize.cpp.o] Error 1
```

--- a/mz_compat.c
+++ b/mz_compat.c
@@ -1093,15 +1093,28 @@ int unzGoToNextFile(unzFile file) {
     return err;
 }
 
-int unzLocateFile(unzFile file, const char *filename, unzFileNameComparer filename_compare_func) {
+#ifdef WIN32
+#  define UNZ_DEFAULT_IGNORE_CASE 1
+#else
+#  define UNZ_DEFAULT_IGNORE_CASE 0
+#endif
+
+int unzLocateFile(unzFile file, const char *filename, unzFileNameCase filename_case) {
     mz_compat *compat = (mz_compat *)file;
     mz_zip_file *file_info = NULL;
     uint64_t preserve_index = 0;
     int32_t err = MZ_OK;
     int32_t result = 0;
+    uint8_t ignore_case = UNZ_DEFAULT_IGNORE_CASE;
 
     if (!compat)
         return UNZ_PARAMERROR;
+
+    if (filename_case == 1) {
+        ignore_case = 0;
+    } else if (filename_case > 1) {
+        ignore_case = 1;
+    }
 
     preserve_index = compat->entry_index;
 
@@ -1111,12 +1124,7 @@ int unzLocateFile(unzFile file, const char *filename, unzFileNameComparer filena
         if (err != MZ_OK)
             break;
 
-        if ((intptr_t)filename_compare_func > 2) {
-            result = filename_compare_func(file, filename, file_info->filename);
-        } else {
-            int32_t case_sensitive = (int32_t)(intptr_t)filename_compare_func;
-            result = mz_path_compare_wc(filename, file_info->filename, !case_sensitive);
-        }
+        result = mz_path_compare_wc(filename, file_info->filename, !ignore_case);
 
         if (result == 0)
             return MZ_OK;

--- a/mz_compat.h
+++ b/mz_compat.h
@@ -308,7 +308,12 @@ typedef struct unz_file_info_s {
 
 /***************************************************************************/
 
-typedef int (*unzFileNameComparer)(unzFile file, const char *filename1, const char *filename2);
+/* Possible values:
+   0 - Uses OS default, e.g. Windows ignores case.
+   1 - Is case sensitive.
+   >= 2 - Ignore case.
+*/
+typedef int unzFileNameCase;
 typedef int (*unzIteratorFunction)(unzFile file);
 typedef int (*unzIteratorFunction2)(unzFile file, unz_file_info64 *pfile_info, char *filename,
     uint16_t filename_size, void *extrafield, uint16_t extrafield_size, char *comment,
@@ -349,7 +354,7 @@ ZEXPORT int     unzGetCurrentFileInfo64(unzFile file, unz_file_info64 * pfile_in
 
 ZEXPORT int     unzGoToFirstFile(unzFile file);
 ZEXPORT int     unzGoToNextFile(unzFile file);
-ZEXPORT int     unzLocateFile(unzFile file, const char *filename, unzFileNameComparer filename_compare_func);
+ZEXPORT int     unzLocateFile(unzFile file, const char *filename, unzFileNameCase filename_case);
 
 ZEXPORT int     unzGetLocalExtrafield(unzFile file, void *buf, unsigned int len);
 

--- a/test/test_compat.cc
+++ b/test/test_compat.cc
@@ -90,7 +90,7 @@ static void test_unzip_compat(unzFile unzip) {
     EXPECT_EQ(global_info64.number_disk_with_CD, 0)
         << "invalid disk with cd 64-bit";
 
-    EXPECT_EQ(err = unzLocateFile(unzip, "test.txt", (unzFileNameComparer)(void *)1), UNZ_OK)
+    EXPECT_EQ(err = unzLocateFile(unzip, "test.txt", 1), UNZ_OK)
         << "cannot locate test file (err: " << err << ")";
 
     EXPECT_EQ(err = unzGoToFirstFile(unzip), UNZ_OK);

